### ISSUE TITLE
IOS wrapper does not use the provided options

### ISSUE THREAD

**Issue by ninio** [zxing-cpp/zxing-cpp#737]
In the demo project of the iOS wrapper, any settings passed to the reader are ignored.

E.g. if you update the options in `viewDidLoad` method of the `ViewController`, they are not used when calling the reader.

Example:
```
class ViewController: UIViewController {
    /* ... */
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.reader.options.tryHarder = false;
         /* ... */
    }
```

**Comment by axxel** [comment]
That sounds embarrassing... ;). I've never seen the iOS code in action myself, so I can't confirm nor contradict your statement. Maybe @benjohnde, @markusfisch, @alexmanzer or @parallaxe can say something about it.

That said, I recently noticed something fishy with regards to the `ReaderOptions` handling myself. Could you maybe try a simple change in the code and see if that helps? I suggest to replace [this line](https://github.com/zxing-cpp/zxing-cpp/blob/b3aff4a98b03e056a244ca385a8221c50d67e352/wrappers/ios/Sources/Wrapper/Reader/ZXIBarcodeReader.mm#L140) with
```ObjectiveC
        Barcodes results = ReadBarcodes(imageView, self.options.cppOpts);
````
And if that does not compile, then there might be some change required in `ZXIReaderOptions.h` to get outside access to [cppOpts](https://github.com/zxing-cpp/zxing-cpp/blob/b3aff4a98b03e056a244ca385a8221c50d67e352/wrappers/ios/Sources/Wrapper/Reader/ZXIReaderOptions.mm#L9). I don't know ObjectiveC.

**Comment by ninio** [comment]
The `ReaderOptions` (`cppOpts`) are once created when initializing the reader (and its options) (`ZXIReaderOptions.mm:16`), but when calling the `reader.read`, a new `ReaderOptions` instance is created (`ZXIBarcodeReader.mm:124`)

I can (try to) create a PR which has the ZXIReaderOptions store only the properties (without a `self.cppOpts`), and do the options ObjC->cpp translations only in the `ReaderOptionsFromZXIReaderOptions`. 

Is this a viable approach, or am I missing something?

**Comment by axxel** [comment]
> The `ReaderOptions` (`cppOpts`) are once created when initializing the reader (and its options) (`ZXIReaderOptions.mm:16`), but when calling the `reader.read`, a new `ReaderOptions` instance is created (`ZXIBarcodeReader.mm:124`)

That is exactly what I meant by 'fishy' and which I believe needs to be changed/removed. The whole `ReaderOptionsFromZXIReaderOptions` method should go away.

> I can (try to) create a PR which has the ZXIReaderOptions store only the properties (without a `self.cppOpts`), and do the options ObjC->cpp translations only in the `ReaderOptionsFromZXIReaderOptions`.

That would also work but will be a larger diff. But if it would end up resulting in less ObjectiveC code overall, I'd prefer that approach.

**Comment by axxel** [comment]
> I can't pass the `self.options.cppOpts` - it's private, so I cannot use it from the BarcodeReader implementation.

That is what I feared, which makes the `cppOpts` member kind of useless. I believe the simplest fix is to make cppOpts public or maybe add a public function that returns it - or even just a pointer to it and use that.

**Comment by ninio** [comment]
> Yes, the `ReaderOptionsFromZXIReaderOptions` method is quite silly now 🙈 This was used before we extended the reader options.
> 
> I've put together a quick [PR](https://github.com/zxing-cpp/zxing-cpp/pull/738) that does use `cppOpts` from `ZXIReaderOptions` directly, and removes the `ReaderOptionsFromZXIReaderOptions` method, which is no longer necessary then.
> 
> But since I am not in the office today, I can't really verify if it's working correctly 😬 It compiles, but that's all I can say right now. But if you're in the middle of trying this out anyway, you could probably check it out and verify it's working as intended.

I'll pull it and see if it works!

**Comment by markusfisch** [comment]
> Is it intentional that the ReaderOptions can only be set upon initializing them?

No, definitely not. This is a bug! 🙈 

I just pushed another [commit](https://github.com/zxing-cpp/zxing-cpp/pull/738/commits/6d2ff5847a2a6627341b923944c632b09c0ce58d) to the PR that makes the setter take effect.

To be honest, I have no idea why this is necessary 🤔

**Comment by ninio** [comment]
Could it be due to `cppOpts`  being a property and not a pointer?

**Comment by axxel** [comment]
Based on that observed behavior, I assume that writing `self.cppOpts` returns a copy of the object, this then gets modified and needs to be assigned back to have an effect. This whole issue here, makes me dislike the ObjectiveC solution even more (see #726). Anyway, I'll merge the PR as is to fix the issue at hand.

@ninio can you confirm that #738 indeed works for you?

**Comment by ninio** [comment]
Yes, the PR fixes this issue!

**Comment by ninio** [comment]
> > Based on that observed behavior, I assume that writing self.cppOpts returns a copy of the object, ...
> 
> Yes, that's exactly the reason!
> 
> I finally found the time to dig a little deeper and it's really quite obvious 🙈 A `@property` in Objective-C is accessed by a (generated) getter and setter function, and because our properties type is `Zxing::ReaderOptions`, not a pointer, the parameter is passed by value, what means it is copied.

Yeah, I thought this might be the reason, but my ObjC++ is quite limited! So thanks for stepping in!

> 
> Sorry for the trouble. We should've stumbled over this a long time ago.

No worries - thank you for being so quick with the fix!
And thanks again for your work on the library!
